File: apps/plant_shop_c-sharp/plant_shop_asp_EF_core/Services/AuthService.cs
using plant_shop_asp_EF_core.Data;
using plant_shop_asp_EF_core.Models;
using plant_shop_asp_EF_core.Utils;
using Microsoft.EntityFrameworkCore;

namespace plant_shop_asp_EF_core.Services
{
    /// <summary>
    /// Service d authentification (validation, inscription, login).
    /// </summary>
    public class AuthService
    {
        private readonly AppDbContext _context;
        private readonly JwtUtil _jwtUtil;

        /// <summary>
        /// Constructeur avec injection du contexte DB et utilitaire JWT.
        /// </summary>
        /// <param name="context">Contexte Entity Framework.</param>
        /// <param name="jwtUtil">Utilitaire de generation JWT.</param>
        public AuthService(AppDbContext context, JwtUtil jwtUtil)
        {
            _context = context;
            _jwtUtil = jwtUtil;
        }

        /// <summary>
        /// Valide les credentials d un utilisateur.
        /// </summary>
        /// <param name="email">Email de l utilisateur.</param>
        /// <param name="password">Mot de passe en clair.</param>
        /// <returns>User si valide, null sinon.</returns>
        public async Task<User?> ValidateUser(string email, string password)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == email);
            if (user != null && PasswordUtil.CheckPassword(password, user.PasswordHash))
            {
                return user;
            }
            return null;
        }

        /// <summary>
        /// Enregistre un nouvel utilisateur.
        /// </summary>
        /// <param name="email">Email unique.</param>
        /// <param name="password">Mot de passe en clair.</param>
        /// <param name="name">Nom optionnel.</param>
        /// <returns>Utilisateur cree.</returns>
        public async Task<User> RegisterUser(string email, string password, string? name)
        {
            var existingUser = await _context.Users.AnyAsync(u => u.Email == email);
            if (existingUser)
            {
                throw new ApplicationException("Email déjà utilisé");
            }

            var user = new User
            {
                Email = email,
                PasswordHash = PasswordUtil.HashPassword(password),
                Name = name,
                IsAdmin = false
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();
            return user;
        }

        /// <summary>
        /// Genere un token JWT pour l utilisateur.
        /// </summary>
        /// <param name="user">Utilisateur authentifie.</param>
        /// <returns>Token JWT.</returns>
        public string Login(User user)
        {
            return _jwtUtil.GenerateToken(user);
        }
    }
}

File: apps/plant_shop_nest/src/app/auth/auth.service.ts
// # Importations
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UsersService } from '../users/users.service';
import * as bcrypt from 'bcryptjs';
import { JwtService } from '@nestjs/jwt';

// # Service d'authentification
@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService
  ) {}

  /**
   * Inscription d'un nouvel utilisateur
   * @param email Email de l'utilisateur
   * @param password Mot de passe en clair
   * @param name Nom de l'utilisateur (optionnel)
   * @return Utilisateur créé (sans mot de passe)
   */
  async register(email: string, password: string, name?: string) {
    const hashed = await bcrypt.hash(password, 10);
    const utilisateur = await this.usersService.create({
      email,
      password: hashed,
      name,
    });

    return {
      id: utilisateur.id,
      email: utilisateur.email,
      name: utilisateur.name,
      admin: utilisateur.admin,
    };
  }

  /**
   * Valide les credentials d'un utilisateur
   * @param email Email de l'utilisateur
   * @param password Mot de passe en clair
   * @return Utilisateur validé
   */
  async validateUser(email: string, password: string) {
    const utilisateur = await this.usersService.findByEmail(email);
    if (!utilisateur) throw new UnauthorizedException('Utilisateur inexistant');
    const valide = await bcrypt.compare(password, utilisateur.password);
    if (!valide) throw new UnauthorizedException('Mot de passe invalide');

    return utilisateur;
  }

  /**
   * Générer un JWT pour un utilisateur
   * @param utilisateur Utilisateur à authentifier
   * @return Token JWT et données utilisateur
   */
  async login(utilisateur: any) {
    const payload = {
      sub: utilisateur.id,
      email: utilisateur.email,
      admin: utilisateur.admin,
      name: utilisateur.name,
    };
    return {
      access_token: this.jwtService.sign(payload),
      user: utilisateur,
    };
  }
}
